### DISCUSSION TITLE
What happens during "task initilization"?

### DISCUSSION THREAD

**Discussion by chbiel** [concourse/concourse#7000] [Help & Support] [answered]
We have a very hard time to speed up our build pipelines.

We clearly understand, why our actual tasks need the time they take, but what we don't understand ist, why some task need much time to initialize.
E.g. we have a task that takes 2m for initialization and 2m for "step".

So our question is:
What is happening in the initialization phase?

I was not able to find any documentation on that.

We also already tried using the experimental tracing feature (https://concourse-ci.org/tracing.html) but that actually also does not give any insight that you cannot see in the WebUI (artifact streaming is visible sometime, but that seems not to be the case here)

**TL;DR**
overlay always does full copies. in our case a docker-in-docker setup makes this worse.
concourse enabled an overlay feature "metacopy" in v7.1.0 what would speed up the initilization time of tasks.

**Comment by taylorsilva** [comment, marked as answer]
How are you deploying Concourse (bosh, helm chart, etc.) and what version are you using?

AFAIK the majority of task initialization time is waiting for baggageclaim to setup all the correct mounts. This means creating Copy-On-Write volumes of the base image for your task and inputs. Do you know what baggageclaim driver you're using  (naive, overlay, btrfs)? We recommend using `overlay` which the worker will try to do if you haven't explicitly picked a driver (via the `detect` driver which is the default choice).

**Reply by chbiel** [reply]
thanks for getting back quickly!

we don't specify a specific driver, so we use the default.
Can  I easily find out which driver concourse picked?

We deploy concourse on ec2 instances without helm etc. so simple `docker run`, newest version 7.2.0

Are there other ways to find out what is happening in detail?
Or are there plans (or maybe undocumented features) to extend the tracing, so that such stuff can be debugged trough that option?

So what I understand from your comment is, that initialization does stuff like
- starting the task image
- creating volumes and mounts
- and such stuff that is necessary to even run e.g. `sh` in the build

Does it also involve streaming artifacts?

EDIT: I fell over https://github.com/concourse/concourse/discussions/6908 that is maybe related, as I would expect, that the docker container uses the binary, so maybe in our case the default is not correct. Or do I missunderstand that disussion?
EDIT2: tried `overlay` explicitly, no improvement 😄

**Reply by taylorsilva** [reply]
You can find out which driver the worker is using by looking in the logs for `baggageclaim.using-driver`.

> Are there other ways to find out what is happening in detail?
AFAIK tracing is your best option. Sadly we don't have any immediate plans to improve tracing or have the bandwidth to do so (we want to though!). 

I'm fairly confident that initialization is all volume setup. This does include streaming from one worker to another. How large are your inputs for some of your tasks? Are they really large, like 1GB+? Same question for the images your tasks are based on, how large are they?

**Reply by konstantin-tw** [reply]
Thank you for the reply!

- Our initial inputs < 100Mb (just source code)
- Our images are  1 - 1,5Gb of size

We see that initialisation time on the same tasks sometimes(rare) is just a few seconds but often it is 1.5-2.5 mins long. It means sometimes (due to some caching? or some combination of conditions?) it can be fast. 
**What can explain fast initialisation time which we see sometimes?** If we understand that we can try to optimise our setup.

EDIT: it looks like the issue with long initialisation are only for the parallel jobs (we have  up to 5 running in parallel) and all of them are taking `git` input from a previous job 
EDIT: the parallelisation was not the root cause; after chaining the jobs sequentially the init time was still high

**Reply by taylorsilva** [reply]
@chbiel someone in a similar discussion mentioned that database IO ended up being the cause for their slow initialization time: https://github.com/concourse/concourse/discussions/6860#discussioncomment-737591

**Reply by konstantin-tw** [reply]
Not sure database IO is an issue for us. We rely on RDS and are not in control of disk there - i think we'll just pay for whatever IOPS it will use, so should not be the limiting factor.
Another observation is that for the jobs that have longer init time we use containers that run docker engine, so it's docker inside docker (for some jobs it is needed to run containerised integration tests), also the size of this image is the biggest 1.6Gb. I'm not sure the size is the problem because on other jobs we use image of 1.3Gb and init time is just 3-7secs

**Reply by taylorsilva** [reply]
Ah, that means those steps running Docker engine have to be privileged containers. If you're using overlay as your baggageclaim driver this can be slow if you're on <7.1.0 of Concourse. In 7.1.0 we [enabled an overlay feature](https://github.com/concourse/concourse/pull/6643) by default, metacopy, if it's detected. This speeds up initialization time of privileged containers so they initialize just as fast a non-privileged containers.

tldr of why privileged containers are slow with baggageclaim: overlay ends up having to copy all volumes. So if your task requires a 1.6Gb image + 3 other volumes, overlay will basically do a `cp` of the image for the task and all volumes.

**Reply by chbiel** [reply]
hey, sorry for not getting back earlier.

As far as I can see, metacopy should be available and we also have concourse 7.2.0 but we don't see any improvements around the initilization.
Are you aware of another way to verify, that the metacopy feature is enabled and used? Or any other idea in which direction we could further dig into?

My last idea is to reduce the image sizes to make it faster, what will also be hard and maybe not fully possible.

It's really a hard task to dig into this deep internals to find out what is going on and why it is slow :D

edit: i actually have another idea, as through the docker in docker stuff, maybe the "inner engine init" is somehow slow...i will dig into that one maybe

**Reply by taylorsilva** [reply]
Concourse checks for the metacopy feature by seeing if the `/sys/module/overlay/parameters/metacopy` file exists. If it exists concourse will use overlay with metacopy on.

> maybe the "inner engine init" is somehow slow

That stuff is part of the task running, not initialization as far as Concourse is concerned. That stuff does not count towards Concourse's initialization of a step.

**Reply by chbiel** [reply]
Thanks, I will now need to find out how to enable it on aws base images ;) getting back when I found it

**Reply by chbiel** [reply]
out of curiosity: did you ever made "metacopy" work on an Amazon Linux machine?

otherwise it seems we are stuck as it seems not to be supported by default :(

**Reply by taylorsilva** [reply]
We don't test on AWS infra. You may be able to enable it yourself by following: https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#metadata-only-copy-up

**Reply by chbiel** [reply]
OK thanks, we will try to figure this out. I will mark it as answered for the moment. Thanks very much :)
